pingcap/tidb
# config, sysvar: add config `instance.enable_ddl` and sysvar `tidb_enable_ddl`

### What problem does this PR solve?


Issue Number: ref #34960 

Problem Summary:

Before, if we want to disable DDL in a node/instance, we have to change the configuration and restart the node. 

### What is changed and how it works?

In this PR we introduce an *instance* scope variable `enable_ddl` and the corresponding config item `instance.enable_ddl`. 

### Check List

Tests 

- [x] Unit test (using previous testing)

Side effects


Documentation


### Manual Test Steps

#### Test 1

1. start a tiup playground with two TiDB instances (port `4000` and `4001`): `tiup playground --db 2 --db.binpath ~/tidb/bin/tidb-server` 
2. start two sessions connecting to two instances separately 
3. using `curl http://127.0.0.1:10081/info/all` to check which instance is the *owner* (assume `4000`)
4. in the session, using `set @@global.tidb_enable_ddl = 0/1` to disable/enable ddl. When disabling ddl in the "owner", another instance should be the owner by checking `curl http://127.0.0.1:10081/info/all`
5. using `curl -X POST http://127.0.0.1:10080/ddl/owner/resign` to let the owner(`4000`) start a new campaign. the instance disabled ddl should not be an owner, and the instance enabled ddl could be an owner.

#### Test 2

1. Start two TiDB servers, say `tidb-a`(DDL owner) and `tidb-b`.
2. Execute on `tidb-a`: `set @@global.tidb_enable_ddl = false;`
3. Stop `tidb-b`, or just send a *resign* signal to `tidb-b`
4. Execute on `tidb-a`: `set @@global.tidb_enable_ddl = true;`
5. Execute on `tidb-a`: `create table t (a int);`

### Release note



Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Previously, TiDB users use the config item `run-ddl` to control whether the TiDB cluster could run DDL.

From now on, users can use the config item `instance.tidb_enable_ddl` to control whether the TiDB instance could run DDL, and the instance system variable `tidb_enable_ddl` to change online.
```
